cmd/pebble/cmd_exec.go

```diff
@@ -0,0 +1,250 @@
+	"timeout":     "Timeout after which to terminate command",
+	"terminal":    "Force pseudo-terminal allocation",
+	"no-terminal": "Disable pseudo-terminal allocation",
+}
+
+var shortExecHelp = "Execute a command and wait for it to finish"
+var longExecHelp = `
+The exec command executes a command via the Pebble API and waits for it to
+finish. Stdin is forwarded, and stdout and stderr are received. By default,
+exec's terminal mode is used if the terminal is a TTY (use -t/--terminal or
+-T/--no-terminal to override).
+
+To avoid confusion, exec options may be separated from the command and its
+arguments using "--", for example:
+
+pebble exec --timeout 10s -- echo -n foo bar
+`
+
+func (cmd *cmdExec) Execute(args []string) error {
+	if cmd.Terminal && cmd.NoTerminal {
+		return errors.New("can't pass -t and -T at the same time")
+	}
+
+	command := append([]string{cmd.Positional.Command}, args...)
+	logger.Debugf("Executing command %q", command)
+
+	// Set up environment variables.
+	env := make(map[string]string)
+	term, ok := os.LookupEnv("TERM")
+	if ok {
+		env["TERM"] = term
+	}
+	for _, kv := range cmd.Env {
+		parts := strings.SplitN(kv, "=", 2)
+		key := parts[0]
+		value := ""
+		if len(parts) == 2 {
+			value = parts[1]
+		}
+		env[key] = value
+	}
+
+	// Send UID if it looks like an integer, otherwise send username.
+	var user string
+	var userID *int
+	uid, err := strconv.Atoi(cmd.User)
+	if err != nil {
+		user = cmd.User
+	} else {
+		userID = &uid
+	}
+
+	// Send GID if it looks like an integer, otherwise send group name.
+	var group string
+	var groupID *int
+	gid, err := strconv.Atoi(cmd.Group)
+	if err != nil {
+		group = cmd.Group
+	} else {
+		groupID = &gid
+	}
+
+	// Specify UseTerminal if -t/--terminal is given, or if both stdin and
+	// stdout are TTYs.
+	stdinIsTerminal := ptyutil.IsTerminal(unix.Stdin)
+	stdoutIsTerminal := ptyutil.IsTerminal(unix.Stdout)
+	var useTerminal bool
+	if cmd.Terminal {
+		useTerminal = true
+	} else if cmd.NoTerminal {
+		useTerminal = false
+	} else {
+		useTerminal = stdinIsTerminal && stdoutIsTerminal
+	}
+
+	// Record terminal state (and restore it before we exit).
+	if useTerminal && stdinIsTerminal {
+		oldState, err := ptyutil.MakeRaw(unix.Stdin)
+		if err != nil {
+			return err
+		}
+		defer ptyutil.Restore(unix.Stdin, oldState)
+	}
+
+	// Grab current terminal dimensions.
+	var width, height int
+	if stdoutIsTerminal {
+		width, height, err = ptyutil.GetSize(unix.Stdout)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Start the command.
+	opts := &client.ExecOptions{
+		Command:     command,
+		Environment: env,
+		WorkingDir:  cmd.WorkingDir,
+		Timeout:     cmd.Timeout,
+		User:        user,
+		UserID:      userID,
+		Group:       group,
+		GroupID:     groupID,
+		UseTerminal: useTerminal,
+		Width:       width,
+		Height:      height,
+		Stdin:       os.Stdin,
+		Stdout:      os.Stdout,
+	}
+	if !useTerminal {
+		opts.Stderr = os.Stderr
```

Comment:
When stderr is split from stdout, the alignment of these two becomes off because the file descriptor is different on the remote side, which is awkward and pretty much always undesired unless stderr is being redirected elsewhere explicitly, in which case alignment is irrelevant.

So the question we must ask is not whether there's a terminal or not, but whether stderr is redirected elsewhere locally so that we can forward that same decision of whether we have a single file descriptor or two file descriptors to the remote end. We can probably do that easily by looking at os.Stdout and os.Stderr.

Reply:
Hmmm, I don't think that's quite it: consider the (common, in my experience) case where you're capturing a program's output. But you still want any error messages to show up visibly in the terminal. For example:

```
pebble exec -- /bin/sh -c 'echo Useful Output; echo ERROR YOU WANT TO SEE! >&2' >out
  # or maybe more realistic, dumping a db schema:
pebble exec -- pg_dump -s mydb >mydb.sql
```

With what you're suggesting the errors would not be shown, but would be mixed in with the output, making it potentially unparseable (unless you explicitly redirected stderr to another file).

However, I agree it shouldn't be directly tied to whether we're using a terminal or not (`useTerminal`). I've changed the logic to, "if you're redirecting either stdout or stderr", split stderr, which I think is what we want.

Reply:
> With what you're suggesting the errors would not be shown, but would be mixed in with the output, making it potentially unparseable (unless you explicitly redirected stderr to another file).

We must be speaking about different things. Whether errors are on file descriptor 1 or file descriptor 2, they are always shown unless you explicitly redirect those somewhere else.

Reply:
I meant that you can't just test "whether stderr is redirected elsewhere locally" to determine this. Because in the case when you just redirect stdout (eg: database dump), you still want stderr (error messages) to be split out and sent to the console.

In any case, I brainstormed about this with the team, and Harry came up with a suggestion that I think does what we actually want: if stdout and stderr refer to the same file/terminal, combine them on the server. The logic is `combineStderr = os.Readlink("/proc/self/fd/1") == os.Readlink("/proc/self/fd/2")`. So if stdout is being redirected to a file but stderr is the terminal, or vice versa, or if both are being redirected to different places, split them. There might be a better way than using `Readlink`, but I haven't found it.

Reply:
Per voice discussion, this is in place now. If we find a better way than `os.Readlink`, we'll use it, but for now this seems to work well - resolving.